File: sdk/node/provider/easynet/pluginexec.js
// Provider-scoped helper for EasyNet-Cli declarative exec plugins.
//
// This module owns the JSON frame details used between `easynet-daemon` and a
// process-backed plugin. Plugin authors should implement handlers over
// SidecarInvocation instead of hand-writing stdin/stdout protocol frames.

import { createInterface } from "node:readline/promises";

export class SidecarProtocolError extends Error {
  constructor(message) {
    super(message);
    this.name = "SidecarProtocolError";
  }
}

export class SidecarInvocation {
  constructor({
    callID,
    caller,
    callee,
    ability,
    subject,
    invocationNonce,
    causalContext,
    args,
    frameType = "invoke",
  }) {
    this.callID = requireString(callID, "call_id");
    this.caller = requireString(caller, "caller");
    this.callee = requireString(callee, "callee");
    this.ability = requireString(ability, "ability");
    this.subject = requireString(subject, "subject");
    this.invocationNonce = requireNonce(invocationNonce, "invocation_nonce");
    this.causalContext = causalContext ?? {};
    this.args = optionalObject(args, "args");
    this.frameType = requireString(frameType, "type");
  }

  static fromFrame(frame) {
    const value = optionalObject(frame, "sidecar request frame", { required: true });
    const frameType = requireString(value.type, "type");
    if (frameType !== "invoke") {
      throw new SidecarProtocolError(
        `exec sidecar expected invoke frame, got ${JSON.stringify(frameType)}`,
      );
    }
    const callID = requireString(value.call_id, "call_id");
    const invocation = optionalObject(value.invocation, "invocation", { required: true });
    return new SidecarInvocation({
      callID,
      caller: invocation.caller,
      callee: invocation.callee,
      ability: invocation.ability,
      subject: invocation.subject,
      invocationNonce: invocation.invocation_nonce,
      causalContext: invocation.causal_context,
      args: invocation.args,
      frameType,
    });
  }
}

export async function serveExecPlugin(handler, options = {}) {
  const input = options.input ?? process.stdin;
  const output = options.output ?? process.stdout;
  let callID = "";
  try {
    const frame = await readFrame(input);
    callID = requireString(frame.call_id, "call_id");
    const invocation = SidecarInvocation.fromFrame(frame);
    const value = await handler(invocation);
    writeFrame(output, {
      type: "result",
      call_id: invocation.callID,
      value,
    });
  } catch (error) {
    writeFrame(output, {
      type: "error",
      call_id: callID,
      message: error instanceof Error ? error.message : String(error),
    });
  }
}

async function readFrame(input) {
  const reader = createInterface({ input, crlfDelay: Infinity });
  try {
    for await (const line of reader) {
      if (line.length === 0) {
        continue;
      }
      try {
        const decoded = JSON.parse(line);
        return optionalObject(decoded, "sidecar request frame", { required: true });
      } catch (error) {
        if (error instanceof SidecarProtocolError) {
          throw error;
        }
        throw new SidecarProtocolError(`invalid sidecar request JSON: ${error.message}`);
      }
    }
    throw new SidecarProtocolError("missing sidecar request frame");
  } finally {
    reader.close();
  }
}

function writeFrame(output, frame) {
  output.write(`${JSON.stringify(frame)}\n`);
}

function requireString(value, field) {
  if (typeof value !== "string" || value.length === 0) {
    throw new SidecarProtocolError(`sidecar frame field ${JSON.stringify(field)} must be a string`);
  }
  return value;
}

function optionalObject(value, field, { required = false } = {}) {
  if (value == null && !required) {
    return {};
  }
  if (typeof value !== "object" || Array.isArray(value)) {
    throw new SidecarProtocolError(`sidecar frame field ${JSON.stringify(field)} must be an object`);
  }
  return value;
}

function requireNonce(value, field) {
  if (!Array.isArray(value) || value.length === 0) {
    throw new SidecarProtocolError(`sidecar frame field ${JSON.stringify(field)} must be a byte array`);
  }
  return value.map((item) => {
    if (!Number.isInteger(item) || item < 0 || item > 255) {
      throw new SidecarProtocolError(
        `sidecar frame field ${JSON.stringify(field)} must contain bytes`,
      );
    }
    return item;
  });
}

File: sdk/python/easynet_sdk/providers/easynet/plugin_exec.py
"""Provider-scoped helper for EasyNet-Cli declarative exec plugins.

This module owns the JSON frame details used between `easynet-daemon` and a
process-backed plugin. Plugin authors should implement a handler over
`SidecarInvocation`; they should not hand-write sidecar protocol frames.
"""

from __future__ import annotations

from collections.abc import Callable, Mapping
from dataclasses import dataclass
import json
import sys
from typing import Any, TextIO


class SidecarProtocolError(ValueError):
    """Raised when a daemon/plugin sidecar frame is structurally invalid."""


@dataclass(frozen=True)
class SidecarInvocation:
    """Typed view of one daemon-admitted plugin invocation."""

    call_id: str
    caller_ura: str
    callee_ura: str
    ability_ura: str
    subject_ura: str
    invocation_nonce: tuple[int, ...]
    causal_context: Any
    args: Mapping[str, Any]
    frame_type: str = "invoke"

    @classmethod
    def from_frame(cls, frame: Mapping[str, Any]) -> "SidecarInvocation":
        """Project a daemon sidecar frame into a handler-facing invocation."""

        frame_type = _required_string(frame, "type")
        if frame_type != "invoke":
            raise SidecarProtocolError(
                f"exec plugin expected invoke frame, got {frame_type!r}"
            )
        call_id = _required_string(frame, "call_id")
        invocation = _required_mapping(frame, "invocation")
        _reject_legacy_tuple_aliases(invocation)
        nonce = _required_nonce(invocation, "invocation_nonce")
        args = _optional_mapping(invocation.get("args"), "args")
        return cls(
            call_id=call_id,
            caller_ura=_required_string(invocation, "caller_ura"),
            callee_ura=_required_string(invocation, "callee_ura"),
            ability_ura=_required_string(invocation, "ability_ura"),
            subject_ura=_required_string(invocation, "subject_ura"),
            invocation_nonce=nonce,
            causal_context=invocation.get("causal_context"),
            args=args,
            frame_type=frame_type,
        )


SidecarHandler = Callable[[SidecarInvocation], Any]


def serve_exec_plugin(
    handler: SidecarHandler,
    *,
    input_stream: TextIO | None = None,
    output_stream: TextIO | None = None,
) -> None:
    """Run one declarative exec plugin invocation.

    The daemon writes one request frame to stdin and expects exactly one
    response frame on stdout. Handler exceptions are converted into daemon
    protocol `error` frames so tracebacks never corrupt stdout framing.
    """

    input_stream = input_stream or sys.stdin
    output_stream = output_stream or sys.stdout
    call_id = ""
    try:
        frame = _read_frame(input_stream)
        call_id = _required_string(frame, "call_id")
        invocation = SidecarInvocation.from_frame(frame)
        value = handler(invocation)
        _write_frame(
            {"type": "result", "call_id": invocation.call_id, "value": value},
            output_stream,
        )
    except Exception as exc:  # noqa: BLE001 - plugin boundary must become a frame.
        _write_frame(
            {"type": "error", "call_id": call_id, "message": str(exc)},
            output_stream,
        )


def _read_frame(input_stream: TextIO) -> Mapping[str, Any]:
    line = input_stream.readline()
    if not line:
        raise SidecarProtocolError("missing sidecar request frame")
    try:
        decoded = json.loads(line)
    except json.JSONDecodeError as exc:
        raise SidecarProtocolError(f"invalid sidecar request JSON: {exc}") from exc
    if not isinstance(decoded, Mapping):
        raise SidecarProtocolError("sidecar request frame must be an object")
    return decoded


def _write_frame(frame: Mapping[str, Any], output_stream: TextIO) -> None:
    output_stream.write(json.dumps(frame, separators=(",", ":")))
    output_stream.write("\n")
    output_stream.flush()


def _required_string(frame: Mapping[str, Any], field: str) -> str:
    value = frame.get(field)
    if not isinstance(value, str) or not value:
        raise SidecarProtocolError(f"sidecar frame field {field!r} must be a string")
    return value


def _reject_legacy_tuple_aliases(frame: Mapping[str, Any]) -> None:
    for legacy, canonical in (
        ("caller", "caller_ura"),
        ("callee", "callee_ura"),
        ("ability", "ability_ura"),
        ("subject", "subject_ura"),
    ):
        if legacy in frame:
            raise SidecarProtocolError(
                f"sidecar frame field {legacy!r} is retired; use {canonical!r}"
            )


def _required_mapping(frame: Mapping[str, Any], field: str) -> Mapping[str, Any]:
    value = frame.get(field)
    return _optional_mapping(value, field, required=True)


def _optional_mapping(
    value: Any,
    field: str,
    *,
    required: bool = False,
) -> Mapping[str, Any]:
    if value is None and not required:
        return {}
    if not isinstance(value, Mapping):
        raise SidecarProtocolError(f"sidecar frame field {field!r} must be an object")
    return value


def _required_nonce(frame: Mapping[str, Any], field: str) -> tuple[int, ...]:
    value = frame.get(field)
    if not isinstance(value, list) or not value:
        raise SidecarProtocolError(
            f"sidecar frame field {field!r} must be a byte array"
        )
    nonce: list[int] = []
    for item in value:
        if not isinstance(item, int) or item < 0 or item > 255:
            raise SidecarProtocolError(
                f"sidecar frame field {field!r} must contain bytes"
            )
        nonce.append(item)
    return tuple(nonce)


__all__ = [
    "SidecarHandler",
    "SidecarInvocation",
    "SidecarProtocolError",
    "serve_exec_plugin",
]
